sdk-typed-search-attributes.md

```diff
@@ -0,0 +1,340 @@
+* We are calling these "typed search attributes" for user getters/setters
+  * It was considered to just overload client options set and workflow upsert to accept old or new type, but that
+    doesn't help getters (can't overload the return and union is unclear to users and hard to deprecate)
+  * It was considered to just move the getter from workflow package/namespace to workflow info object and vice versa,
+    but this became confusing (in Go they are already on info, in Java they are on workflow context not info, etc)
+  * It was considered to make this just "search attributes" and find ways per language to cheat that (e.g. in Java
+    `Workflow.getSearchAttributes` would be deprecated in favor of `Workflow.searchAttributes`)
+    * This becomes really difficult and inconsistent across languages
+  * Other names considered were:
+    * "workflow search attributes" - doesn't work for schedules
+    * "search attrs" - hard to justify abbreviation
+    * "attributes" - not acceptable to how Temporal uses the "search attributes" term everywhere
+* All existing forms of search attribute getting and setting are hereby deprecated
+* Client options will have ability to get/set this collection as "typed search attributes"
+* Workflows can get a read-only form of this collection (however lang expresses that best) as "typed search attributes"
+  getter at the top of the workflow adjacent to where getting "info" is today
+* "upsert search attributes" is deprecated
+  * We could have overloaded this in all languages but Go. Maybe we should?
+* "update typed search attributes" is made available on the workflow level where "upsert" already exists today
+  * Considered calling "upsert" but that term does not take deletion into account
+    * Should we consider going back on this and representing null/None as a deletion? Problem here is we didn't expect
+      the collection to be able to store nulls.
+    * Maybe we should call it "upsert" since that's what the event name in UI is?
+  * This accepts a vararg `SearchAttributeUpdate` that is either a "set" or a "delete"
+    * We accept this is ugly, but there is no better way to represent deletes
+    * We considered a mutable `SearchAttributes` collection but individual key/value mutations would be too hard to
+      reason about if buffered/batched and if single upsert commands people would ask for bulk as soon as individual
+      upsert events appeared
+      * If we change back to representing deletes as null/None, we can accept mutable form of this collection
+    * Each `SearchAttributeKey` has a `valueSet(value)` and a `valueUnset()` that return `SearchAttributeUpdate`s
+      * So Java may look like `Workflow.updateSearchAttributes(MY_ATTR1.valueSet("foo"), MY_ATTR2.valueUnset())`
+      * Ok that we call it "unset" instead of "delete"?
+      * Ok that we call it `valueSet` and `valueDelete` instead of just `set` and `delete`?
```

Comment:
I think these are fine but I'd just switch it around to be `setValue` and `unsetValue`

Reply:
The reason I put the noun first is because the verb makes you think you're doing something instead of just building an update